libyul/backends/evm/EVMBuiltins.h

```diff
@@ -0,0 +1,118 @@
+	/// Function to generate code for the given function call and append it to the abstract
+	/// assembly. Expects all non-literal arguments of the call to be on stack in reverse order
+	/// (i.e. right-most argument pushed first).
+	/// Expects the caller to set the source location.
+	std::function<void(FunctionCall const&, AbstractAssembly&, BuiltinContext&)> generateCode;
+};
+
+/// Collection of all possible EVM builtin functions.
+/// Each builtin can have one (or multiple) scopes, which define whether, e.g., it requires object access.
+/// Using this class as single source of truth for builtin functions makes sure that these are consistent over
+/// EVM dialects. If the order were to depend on the EVM dialect - which can easily happen using conditionals -,
+/// different dialects' builtin handles become inherently incompatible.
+class EVMBuiltins
+{
+	static std::size_t constexpr instructionBit = 0;
+	static std::size_t constexpr replacedInstructionBit = 1;
+	static std::size_t constexpr objectAccessBit = 2;
+	static std::size_t constexpr requiresEOFBit = 3;
+	static std::size_t constexpr requiresNonEOFBit = 4;
+public:
+	struct Scopes
+	{
+		/// whether the corresponding evm builtin function is an instruction builtin
+		bool instruction() const { return value.test(instructionBit); }
+		/// whether the corresponding evm builtin has been replaced by another builtin, ie, should be skipped
+		bool replaced() const { return value.test(replacedInstructionBit); }
+		/// if true, the evm builtin function is only valid when object access is given
+		bool requiresObjectAccess() const { return value.test(objectAccessBit); }
+		/// if true, the evm builtin function is only valid if EOF is enabled
+		bool requiresEOF() const { return value.test(requiresEOFBit); }
+		/// if true, the evm builtin function is only valid if EOF is not enabled
+		bool requiresNonEOF() const { return value.test(requiresNonEOFBit); }
+
+		Scopes operator|(Scopes const& _other) const
+		{
+			Scopes result = *this;
+			result |= _other;
+			return result;
+		}
+
+		Scopes& operator|=(Scopes const& _other)
+		{
+			value |= _other.value;
+			return *this;
+		}
+
+		std::bitset<5> value;
+	};
+
+	EVMBuiltins();
+
+	std::vector<std::tuple<Scopes, BuiltinFunctionForEVM>> const& functions() const { return m_functions; }
+
+	/// Creates a verbatim builtin function. These are not part of the usual builtin functions collection and
+	/// must be cached in the dialect creating them.
+	static BuiltinFunctionForEVM createVerbatimFunction(size_t _arguments, size_t _returnVariables);
+	static SideEffects sideEffectsOfInstruction(evmasm::Instruction _instruction);
+private:
+	static Scopes constexpr instruction{1 << instructionBit};
+	static Scopes constexpr replaced{1 << replacedInstructionBit};
+	static Scopes constexpr objectAccess{1 << objectAccessBit};
+	static Scopes constexpr requiresEOF{1 << requiresEOFBit};
+	static Scopes constexpr requiresNonEOF{1 << requiresNonEOFBit};
```

Comment:
```suggestion
	static Scopes constexpr instructionScope{1 << instructionBit};
	static Scopes constexpr nonInstructionScope{1 << replacedInstructionBit};
	static Scopes constexpr objectAccessScope{1 << objectAccessBit};
	static Scopes constexpr eofScope{1 << requiresEOFBit};
	static Scopes constexpr nonEOFScope{1 << requiresNonEOFBit};
```

Alternatively, might be enough to place them inside `Scopes` without the suffix so that you can refer to them as e.g. `Scope::eof`, but that would conflict with functions you already have there.

Reply:
I prefer it without the `Scope` suffix tbh. The type already tells the rest of the story.